src/StudyEntities/Candidate/IdentifierGenerator.php

```diff
@@ -0,0 +1,144 @@
+<?php declare(strict_types=1);
+
+namespace LORIS\StudyEntities\Candidate;
+
+abstract class IdentifierGenerator
+{
+    protected $generationMethod;
+    protected $alphabet = array();
+    protected $length;
+    protected $minValue;
+    protected $maxValue;
+    protected $prefix    = '';
+
+    /**
+    * Generates a new unused identifier to represent a Candidate in LORIS.
+    *
+    * @return string The new identifier
+    */
+    protected function createNewID(): string
+    {
+        $this->checkIDRangeFull();
+
+        $id = '';
+        switch ($this->generationMethod) {
+        case 'sequential':
+            $id = $this->padID($this->generateSequentialID());
+            break;
+        case 'random':
+            $id = $this->padID($this->generateRandomID());
+            // If the Random ID is already in use or is outside of the range 
+            // defined by minValue and maxValue, increment it by using the
+            // generateSequentialID function with the previous ID as a starting
+            // point.
+            if (in_array($id, $this->getExistingIDs(), true)
+                || strcmp(strval($id), strval($this->minValue)) < 0
+                || strcmp(strval($id), strval($this->maxValue)) > 0
+            ) {
+                $id = $this->generateSequentialID($id);
+            }
+            break;
+        }
+
+        // Return padded value with appropriate prefix
+        return $this->prefix . $id;
+    }
+
+    protected function generateSequentialID(string $id = ''): string
+    {
+        // If this is the first ID ever created, return the minimum value.
+        if (count($this->getExistingIDs()) < 1) {
+            return $this->minValue;
+        }
+        // Create the new ID by incrementing the value of the $id parameter OR
+        // by incrementing the highest existing ID.
+        $newID = !empty($id) ? $id : max($this->getExistingIDs());
+        // Increment newID until we find an unused value within the boundaries
+        // of $min and $max.
+        do {
+            // Make sure $newID is greater than or equal to minimum and less
+            // than or equal to maximum.
+            if (strcmp(strval($newID++), strval($this->minValue)) < 0) {
+                // If newID is less than the minimum, increment again.
+                continue;
+            }
+            if (strcmp(strval($newID), strval($this->maxValue)) > 0) {
+                // If newID has exceeded the maximum, wrap back around to the
+                // minimum possible value. We know that there is a free ID
+                // available somewhere in the range of possible IDs or else
+                // _generateID above will have thrown an exception. This allows
```

Comment:
What if it was filled in between the start of the loop and here in a different request?

I'm also not convinced that we should be re-using IDs if somehow one got deleted in the middle, I could foresee that causing a lot of ambiguity in the data.

Reply:
I can add another check for existing IDs in this loop.

I can throw an exception if the maximum ID is equal to the max value as defined by the range instead of looping around but this might add a weird edge case if the most-recent Candidate gets deleted. I'm not sure how to enforce that its ID won't get re-used as there will be no record of it in the DB.

Reply:
They should never be deleted anyways, they should just be turned inactive. I think it's safest to just throw an exception if it's greater than the maximum saying you've run out of IDs.